### DISCUSSION TITLE
Is it possible to run in GPU mode while in WASM mode, or is native mode required for GPU?

### DISCUSSION THREAD

**Discussion by pepper-lees** [nextcloud/recognize#801] [Q&A] [answered]
I'm using Nextcloud AIO, and the maintainer said that it's currently not possible to install libtensorflow in an AIO instance - see my question here: https://github.com/nextcloud/all-in-one/discussions/2330

Is it possible for GPU mode to be enabled when in WASM mode? It's a bit unclear, since I see no error or info messages about GPU mode. If GPU mode is **not possible** while WASM mode is enabled, I'd expect the GPU mode switch to be disabled when WASM is enabled; but likewise, I'd also expect some sort of info or error messages about whether GPU mode is working. I installed all the dependencies on my Windows system, but I have no clue whether they're accessible to the nextcloud instance.

![image](https://user-images.githubusercontent.com/11487870/232118728-1bbee318-6df9-46da-961e-be9e421c9a9c.png)

**Comment by marcelklehr** [comment, marked as answer]
Yes, WASM mode and GPU mode exclude one another. Basically, you can run in WASM, natively on CPU, or on GPU. I will fix this in the admin settings UI. Thank you for reporting! :)

**Comment by tetebueno** [comment]
Hey there, just bump into this: my processor doesn't support AVX, so WASM is enabled; I also have a GPU available with everything mentioned in the wiki installed in order to use it for processing. Is there a way I can use the GPU?

**Reply by tetebueno** [reply]
Just to clarify, this is the output I get when testing capabilities (only WASM test script exits successfully):

```console
www-data@aed84ee21bde:~$ ./bin/node src/test_wasmtensorflow.js
www-data@aed84ee21bde:~$ ./bin/node src/test_gputensorflow.js
Illegal instruction (core dumped)
www-data@aed84ee21bde:~$ ./bin/node src/test_libtensorflow.js
Illegal instruction (core dumped)
```

**Reply by tetebueno** [reply]
Just took a logic lesson and relized the meaning of this:

> WASM mode and GPU mode exclude one another

So I won't be bothering here anymore.

Cheers.
